src/simulator/halo_simulator.py

```diff
@@ -228,12 +229,47 @@ def simulated_venn_diagram_reach_by_spend(
               satisfying the request.
             max_frequency:  The maximum frequency for which to report reach.
         Returns:
-            A list of ReachPoint. Each reach point represents the mapping from 
+            A list of ReachPoint. Each reach point represents the mapping from
             the spends of a subset of publishers to the differentially private
             estimate of the number of people reached in this subset.
         """
         raise NotImplementedError()
 
+    def _aggregate_reach_in_venn_diagram_regions(
+        self, pub_ids: List[int], venn_diagram_regions: Dict[int, List]
+    ) -> int:
+        """Returns an aggregated reach from Venn diagram regions given
+        publisher IDs.
+
+        With the Venn diagram regions, we aggregate reach from the regions
```

Comment:
I don't really understand this paragraph.

Reply:
The first sentence explained that we only pick the regions that contain at least one of the given publishers for reach aggregation. The second sentence explained that the binary representation of the index of a region in `venn_diagram_regions` can be seen as a bit map showing which publishers IDs belong to this region. This explains the code
```python
targeted_pub_repr = sum(1 << pub_id for pub_id in pub_ids)
```
and
```python
if r & targeted_pub_repr
```

Reply:
Would it be better to rephrase the paragraph as following?
`With the Venn diagram regions, we aggregate reach from the regions where contain at least one of the given publishers. Note that the binary representation of the index of a region represents the formation of publisher IDs in that region.`

Reply:
Add @jiayu-google to the thread to avoid separate discussions.

Add Jiayu's comment here:
> Would you consider rephrasing this sentence. An ideas is as follows.
To obtain the union reach of the given subset of publishers, we sum up the reach from the regions of which the binary representation equals one at the digit corresponding to any of the given publishers. For example, .....

Reply:
Incorporate Jiayu's suggestion:

```
To obtain the union reach of the given subset of publishers, we sum up the 
reaches from the regions in which all users are reached by at least one of the 
given publishers. Note that the binary representation of the index of a region 
represents the formation of publisher IDs in that region. 

For example, given a subset of publisher ids, {0}, out of the whole set {0, 1, 2}, 
the reaches in the following regions will be summed up:

    region with index = 1 = bin('001'): reached by pub_id-0
    region with index = 3 = bin('011'): reached by pub_id-0 and 1
    region with index = 5 = bin('101'): reached by pub_id-0 and 2
    region with index = 7 = bin('111'): reached by pub_id-0, 1, and 2
```

@jiayu-google @matthewclegg PLMKWYT. Thanks.

Reply:
Thanks Rieman. Looks good except two suggestion as follows.

"regions in which all users are reached by at least one of the given publishers." --- it does not sound precise to me. How about "regions which belong to at least one of the given publisher"?

Also, may replace 'reached by' with 'belongs to'.  Like
      region with index = 1 = bin('001'): belongs to pub_id-0.
We probably don't say 'a region is reached by a pub.'

Reply:
Yes. I was not sure about the correct wording when I was writing it 😅. Thank you so much for the suggestions!

Reply:
Also, add the word 'primitive' before 'regions', where appropriate.

Reply:
Done.